src/Core/DatabaselessKernel.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace SilverStripe\Core;
+
+/**
+ * Boot a kernel without requiring a database connection.
+ * This is a workaround for the lack of composition in the boot stages
+ * of CoreKernel, as well as for the framework's misguided assumptions
+ * around the availability of a database for every execution path.
+ *
+ * @internal
+ */
+class DatabaselessKernel extends CoreKernel
```

Comment:
Is there any hope of adding a KernelInterface and have both DatabaselessKernel + CoreKernel implementing that instead of DatabaselessKernel extending CoreKernel - it's really weird extending something and taking something away from it

Reply:
Agree, but we can't add an interface in a minor release. That's why we're subclassing.

Reply:
I think it's doable if you did this:

```
Kernel (Interface)
|- SomeKernel (New class - move bunch of stuff from CoreKernel to here)
  |- CoreKernel (Include database connection methods)
  |- DatabaselessKernel
```

That would mean that CoreKernel keeps the same signature and functionality

Reply:
So we're going from adding one new `@internal` class to solve an edge case to three new classes and one new interface?

I'm just concerned that the motivation behind this was to keep it as simple and low-impact as possible, and now we're just expanding the maintenance footprint. I agree it's a better architecture, but I just don't want to lose sight over why were all here.

Reply:
We're adding one new abstract class + one new class, as opposed to one new class.  We'll have the same number of methods, so it's literally just one extra file to make things tidier which IMO will reduce the maintenance footprint..  Also as Michal mentioned above the DatabaselessKernel could be used for speedier functional testing as well.

Reply:
Ah, right, I didn't see that the interface already existed. OK, yeah this makes more sense now. Thanks.

Reply:
> move bunch of stuff from CoreKernel to here

I think this is the only bit I'm not sold on. What "stuff" we move is kind of important. In this case, moving all the database-related methods out of the base class and into the CoreKernel would achieve what we need, but it's a bit arbitrary to draw the line at database concerns. If the point of this is abstraction and composability, then why not allow for configless or manifestless kernels as well? If the base class is going to be database agnostic, I see no reason why it can't be agnostic to all the other layers as well.

Here's one option:

* `BaseKernel` becomes a collection of services that can be used to boot the kitchen sink of application layers, e.g. `setThemeResourceLoader`, `setInjectorLoader`, etc..

* `boot()` method remains abstract

* All `bootXXX` protected methods, e.g. `bootConfigs()` defined in the concretion

* To maintain a horizontal architecture, we move `bootXXX()` methods to traits

* Implementations can pick and choose which boots they want. That way CoreKernel and Databaseless kernel can both use the `ManifestBoot` and `ConfigBoot` trait, for example.

* `bootPHP` would be the exception. A non-PHP boot isn't plausible. :-)

Reply:
Trait is probably overkill, though I know you love your composability :-)  Maybe a middle ground is just leave all the stuff you'd move to traits in BaseKernel and just use the abstract boot() method to choose what get booted

I'm not overly fussed how you split it though

Reply:
It's too complicated. There's too much coupling in those boot methods, and it's probably not realistic to boot without a manifest.

A few leaky abstractions here and there, too, e.g. `isFlushed()` not being part of the interface, yet expected in the public consumption of the API. Annoying.

Have refactored.